File: 23.java
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<Integer> heap = new PriorityQueue<>();
        for(ListNode node : lists) {
            while(node != null) {
                heap.add(node.val);
                node = node.next;
            }
        }

        ListNode head = new ListNode();
        ListNode node = head;
        while(!heap.isEmpty()) {
            node.next = new ListNode(heap.poll());
            node = node.next;
        }
        return head.next;
    }
}

File: 23.py
import heapq

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap = []

        for node in lists:
            while node:
                heapq.heappush(heap, node.val)
                node = node.next

        head = ListNode(0)
        current = head

        while heap:
            current.next = ListNode(heapq.heappop(heap))
            current = current.next

        return head.next
